README.md
[![GitHub CI build status](https://github.com/WizardMac/ReadStat/workflows/build/badge.svg)](https://github.com/WizardMac/ReadStat/actions)
[![Appveyor build status](https://ci.appveyor.com/api/projects/status/76ctatpy3grlrd9x/branch/master?svg=true)](https://ci.appveyor.com/project/evanmiller/readstat/branch/master)
[![codecov](https://codecov.io/gh/WizardMac/ReadStat/branch/master/graph/badge.svg)](https://codecov.io/gh/WizardMac/ReadStat)
[![Fuzzing Status](https://oss-fuzz-build-logs.storage.googleapis.com/badges/readstat.svg)](https://bugs.chromium.org/p/oss-fuzz/issues/list?sort=-opened&can=1&q=proj:readstat)

ReadStat: Read (and write) data sets from SAS, Stata, and SPSS
==

Originally developed for [Wizard](https://www.wizardmac.com/), ReadStat is a
command-line tool and MIT-licensed C library for reading files from popular
stats packages. Supported data formats include:

* SAS: SAS7BDAT (binary file) and XPORT (transport file)
* Stata: DTA (binary file) versions 104-119
* SPSS: POR (portable file), SAV (binary file), and ZSAV (compressed binary)

Supported metadata formats include:

* SAS: SAS7BCAT (catalog file) and .sas (command file)
* Stata: .dct (dictionary file)
* SPSS: .sps (command file)

There is also write support for all the data formats, but not the metadata
formats. *The produced SAS7BDAT files still cannot be read by SAS*, but feel
free to contribute your binary-format expertise here.

For reading in R data files, please see the related
[librdata](https://github.com/WizardMac/librdata) project.

Installation on Unix / macOS
--

Grab the latest [release](https://github.com/WizardMac/ReadStat/releases) and
then proceed as usual:

    ./configure
    make
    sudo make install

If you're cloning the repository, first make sure you have autotools installed,
and then run `./autogen.sh` to generate the configure file.

If you're on Mac and see errors about `AM_ICONV` when you run `./autogen.sh`,
you'll need to install [gettext](https://www.gnu.org/software/gettext/).

Installation on Windows
--

ReadStat now includes a Microsoft Visual Studio project file that includes
build targets for the library and tests. See the [VS17](./VS17) folder in
the downloaded release for a "one-click" Windows build.

Alternatively, you can build ReadStat on the command line using an
[msys2](https://msys2.github.io/) environment. After installing msys2,
download some other packages:

    pacman -S autoconf automake libtool make mingw-w64-x86_64-toolchain mingw-w64-x86_64-cmake mingw-w64-x86_64-libiconv

Then start a MINGW command line (not the msys2 prompt!) and follow the UNIX
install instructions above for this package.

Language Bindings
--

* Julia: [ReadStat.jl](https://github.com/queryverse/ReadStat.jl)
* Perl 6: [ReadStat.pm6](https://github.com/WizardMac/ReadStat.pm6)
* Python: [pyreadstat](https://github.com/Roche/pyreadstat)
* R: [haven](https://github.com/tidyverse/haven)


Docker
--

A dockerized version is available [here](https://github.com/jbn/readstat)



Command-line Usage
--

Standard usage:

    readstat [-f] <input file> <output file>

Where:

* `<input file>` ends with `.dta`, `.por`, `.sav`, `.sas7bdat`, or `.xpt`and
* `<output file>` ends with `.dta`, `.por`, `.sav`, `.sas7bdat`, `.xpt` or `.csv`

If [libxlsxwriter](http://libxlsxwriter.github.io) is found at compile-time, an
XLSX file (ending in `.xlsx`) can be written instead.

If zlib is found at compile-time, compressed SPSS files (`.zsav`) can be read
and written as well.

Use the `-f` option to overwrite an existing output file.

If you have a plain-text file described by a Stata dictionary file, a SAS
command file, or an SPSS command file, a second invocation style is supported:

    readstat <input file> <dictionary file> <output file>

Where:

* `<input file>` can be anything
* `<dictionary file>` ends with `.dct`, `.sas`, or `.sps`
* `<output file>` ends with `.dta`, `.por`, `.sav`, `.xpt`, or `.csv`

If you have a SAS catalog file containing the data set's value labels, you
can use the same invocation:

    readstat <input file> <catalog file> <output file>

Except where:

* `<input file>` ends with `.sas7bdat`
* `<catalog file>` ends with `.sas7bcat`
* `<output file>` ends with `.dta`, `.por`, `.sav`, `.xpt`, or `.csv`

If the file conversion succeeds, ReadStat will report the number of rows and
variables converted, e.g.

    Converted 111 variables and 160851 rows in 12.36 seconds

At the moment value labels are supported, but the finer nuances of converting
format strings (e.g. `%8.2g`) are not.

Command-line Usage with CSV input
--

A prerequisite for CSV input is that the [libcsv](https://github.com/rgamble/libcsv)
library is found at compile time.

CSV input is supported together with a metadata file describing the data:

    readstat <input file.csv> <input metadata.json> <output file>

The `<output file>` should end with `.dta`, `.sav`, or `.csv`.

The `<input file.csv>` is a regular CSV file.

The `<input metadata.json>` is a JSON file describing column types, value
labels and missing values. The easiest way to create such a metadata file is to
use the provided `extract_metadata` program on an existing file:

    $ extract_metadata <input file.(dta|sav|sas7bcat)>

The schema of this JSON file is fully described in
[variablemetadata_schema.json](variablemetadata_schema.json) using [JSON
Schema](http://json-schema.org/).

The following is an example of a valid metadata file:

    {
        "type": "SPSS",
        "variables": [
            {
                "type": "NUMERIC",
                "name": "citizenship",
                "label": "Citizenship of respondent",
                "categories": [
                    {
                        "code": 1,
                        "label": "Afghanistan"
                    },
                    {
                        "code": 2,
                        "label": "Albania"
                    },
                    {
                        "code": 98,
                        "label": "No answer"
                    },
                    {
                        "code": 99,
                        "label": "Not applicable"
                    }
                ],
                "missing": {
                    "type": "DISCRETE",
                    "values": [
                        98,
                        99
                    ]
                }
            }
        ]
    }

Here the column `citizenship` is a numeric column with four possible values `1`, `2`, `98`, and `99`.
`1` has the label `Afghanistan`, `2` has `Albania`, `98` has `No answer` and `99` has `Not applicable`.
`98` and `99` are defined as missing values.

Other column types are `STRING` and `DATE`. 
All values in `DATE` columns are expected to conform to [ISO 8601 date](https://en.wikipedia.org/wiki/ISO_8601).
Here is an example of `DATE` metadata:

    {
        "type": "SPSS",
        "variables": [
            {
                "type": "DATE",
                "name": "startdate",
                "label": "Start date",
                "categories": [
                    {
                        "code": "6666-01-01",
                        "label": "no date available"
                    }
                ],
                "missing": {
                    "type": "DISCRETE",
                    "values": [
                        "6666-01-01",
                        "9999-01-01"
                    ]
                }
            }
        ]
    }

Value labels are supported for `DATE`.

The last column type is `STRING`:

    {
        "type": "SPSS",
        "variables": [
            {
                "type": "STRING",
                "name": "somestring",
                "label": "Label of column",
                "missing": {
                    "type": "DISCRETE",
                    "values": [
                        "NA",
                        "N/A"
                    ]
                }
            }
        ]
    }

Value labels are not supported for `STRING`.

Library Usage: Reading Files
--

The ReadStat API is callback-based. It uses very little memory, and is suitable
for programs with progress bars.  ReadStat uses
[iconv](https://en.wikipedia.org/wiki/Iconv) to automatically transcode
text data into UTF-8, so you don't have to worry about character encodings. 

See src/readstat.h for the complete API. In general you'll provide a filename
and a set of optional callback functions for handling various information and
data found in the file. It's up to the user to store this information in an
appropriate data structure. If a context pointer is passed to the parse_* functions,
it will be made available to the various callback functions.

Callback functions should return `READSTAT_HANDLER_OK` (zero) on success.
Returning `READSTAT_HANDLER_ABORT` will abort the parsing process.

Example: Return the number of records in a DTA file.

```c
#include "readstat.h"

int handle_metadata(readstat_metadata_t *metadata, void *ctx) {
    int *my_count = (int *)ctx;

    *my_count = readstat_get_row_count(metadata);

    return READSTAT_HANDLER_OK;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <filename>\n", argv[0]);
        return 1;
    }
    int my_count = 0;
    readstat_error_t error = READSTAT_OK;
    readstat_parser_t *parser = readstat_parser_init();
    readstat_set_metadata_handler(parser, &handle_metadata);

    error = readstat_parse_dta(parser, argv[1], &my_count);

    readstat_parser_free(parser);

    if (error != READSTAT_OK) {
        printf("Error processing %s: %d\n", argv[1], error);
        return 1;
    }
    printf("Found %d records\n", my_count);
    return 0;
}
```

Example: Convert a DTA to a tab-separated file.

```c
#include "readstat.h"

int handle_metadata(readstat_metadata_t *metadata, void *ctx) {
    int *my_var_count = (int *)ctx;
    
    *my_var_count = readstat_get_var_count(metadata);

    return READSTAT_HANDLER_OK;
}

int handle_variable(int index, readstat_variable_t *variable, 
    const char *val_labels, void *ctx) {
    int *my_var_count = (int *)ctx;

    printf("%s", readstat_variable_get_name(variable));
    if (index == *my_var_count - 1) {
        printf("\n");
    } else {
        printf("\t");
    }

    return READSTAT_HANDLER_OK;
}

int handle_value(int obs_index, readstat_variable_t *variable, readstat_value_t value, void *ctx) {
    int *my_var_count = (int *)ctx;
    int var_index = readstat_variable_get_index(variable);
    readstat_type_t type = readstat_value_type(value);
    if (!readstat_value_is_system_missing(value)) {
        if (type == READSTAT_TYPE_STRING) {
            printf("%s", readstat_string_value(value));
        } else if (type == READSTAT_TYPE_INT8) {
            printf("%hhd", readstat_int8_value(value));
        } else if (type == READSTAT_TYPE_INT16) {
            printf("%hd", readstat_int16_value(value));
        } else if (type == READSTAT_TYPE_INT32) {
            printf("%d", readstat_int32_value(value));
        } else if (type == READSTAT_TYPE_FLOAT) {
            printf("%f", readstat_float_value(value));
        } else if (type == READSTAT_TYPE_DOUBLE) {
            printf("%lf", readstat_double_value(value));
        }
    }
    if (var_index == *my_var_count - 1) {
        printf("\n");
    } else {
        printf("\t");
    }

    return READSTAT_HANDLER_OK;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <filename>\n", argv[0]);
        return 1;
    }
    int my_var_count = 0;
    readstat_error_t error = READSTAT_OK;
    readstat_parser_t *parser = readstat_parser_init();
    readstat_set_metadata_handler(parser, &handle_metadata);
    readstat_set_variable_handler(parser, &handle_variable);
    readstat_set_value_handler(parser, &handle_value);

    error = readstat_parse_dta(parser, argv[1], &my_var_count);

    readstat_parser_free(parser);

    if (error != READSTAT_OK) {
        printf("Error processing %s: %d\n", argv[1], error);
        return 1;
    }
    return 0;
}
```

Library Usage: Writing Files
--

ReadStat can write data sets to a number of file formats, and uses largely the
same API for each of them. Files are written incrementally, with the header
written first, followed by individual rows of data, and ending with some kind
of trailer. (So the full data file never resides in memory.) Unlike like the
callback-based API for reading files, the writer API consists of function that
the developer must call in a particular order. The complete API can be found in
[readstat.h](./src/readstat.h).

Basic usage:

```c
#include "readstat.h"

/* A callback for writing bytes to your file descriptor of choice */
/* The ctx argument comes from the readstat_begin_writing_xxx function */
static ssize_t write_bytes(const void *data, size_t len, void *ctx) {
    int fd = *(int *)ctx;
    return write(fd, data, len);
}

int main(int argc, char *argv[]) {
    readstat_writer_t *writer = readstat_writer_init();
    readstat_set_data_writer(writer, &write_bytes);
    readstat_writer_set_file_label(writer, "My data set");

    int row_count = 1;

    readstat_variable_t *variable = readstat_add_variable(writer, "Var1", READSTAT_TYPE_DOUBLE, 0);
    readstat_variable_set_label(variable, "First variable");

    /* Call one of:
     *   readstat_begin_writing_dta
     *   readstat_begin_writing_por
     *   readstat_begin_writing_sas7bdat
     *   readstat_begin_writing_sav
     *   readstat_begin_writing_xport
     */

    int fd = open("something.dta", O_CREAT | O_WRONLY);
    readstat_begin_writing_dta(writer, &fd, row_count);

    int i;
    for (i=0; i<row_count; i++) {
        readstat_begin_row(writer);
        readstat_insert_double_value(writer, variable, 1.0 * i);
        readstat_end_row(writer);
    }

    readstat_end_writing(writer);
    readstat_writer_free(writer);
    close(fd);

    return 0;
}
```

Fuzz Testing
--

To assist in fuzz testing, ReadStat ships with target files designed to work
with [libFuzzer](http://llvm.org/docs/LibFuzzer.html). Clang 6 or later is
required.

1. `./configure --enable-fuzz-testing` turns on useful sanitizer and sanitizer-coverage flags
1. `make` will create a new binary called `generate_corpus`. Running this
   program will use the ReadStat test suite to create a corpus of test files in
   `corpus/`. There is a subdirectory for each sub-format (`dta104`, `dta105`,
   etc.). Currently a total of 468 files are created.
1. If fuzz-testing has been enabled, `make` will also create fourteen fuzzer
   targets, one for each of seven file formats, five for internally used
   grammars, and two fuzzers for testing the compression routines.
   * `fuzz_format_dta`
   * `fuzz_format_por`
   * `fuzz_format_sas7bcat`
   * `fuzz_format_sas7bdat`
   * `fuzz_format_sav`
   * `fuzz_format_xport`
   * `fuzz_format_stata_dictionary`
   * `fuzz_grammar_dta_timestamp`
   * `fuzz_grammar_por_double`
   * `fuzz_grammar_sav_date`
   * `fuzz_grammar_sav_time`
   * `fuzz_grammar_spss_format`
   * `fuzz_compression_sas_rle`
   * `fuzz_compression_sav`

For best results, each sub-directory of the corpus should be passed to the relevant fuzzer, e.g.:

* `./fuzz_format_dta corpus/dta104`
* `./fuzz_format_dta corpus/dta110`
* ...
* `./fuzz_format_sav corpus/sav`
* `./fuzz_format_sav corpus/zsav`
* `./fuzz_format_xport corpus/xpt5`
* `./fuzz_format_xport corpus/xpt8`

Finally, the compression fuzzers can be invoked without a corpus:

* `./fuzz_compression_sas_rle`
* `./fuzz_compression_sav`




NEWS
New in 1.1.8:

    * XPT reader/writer: Improved support for format strings #257 #258
    * DTA writer: Fix off-by-one error in v,o indexing for string refs #270
    * SAV/DTA writers: Improved checking of non-ASCII characters #256
    * SAS7BDAT reader: Fix use-after-free error #273
    * Build: Link to libm on GNU systems #255
    * SAS commands: Support more syntax
    * SPSS commands: Make file names optional

New in 1.1.7:

    * SAV writer: Permit missing ranges for integer variables #251
    * SAV writer: Fix format-truncation warning #247
    * DTA reader/writer: Improved support for string refs on older platforms
    * DTA reader: Fix incorrect ordering in StrL comparison functions #248
    * SAS7BDAT writer: Improved support for tagged missing values on big-endian platforms
    * SAS readers: Allow header sizes up to 16MiB #249
    * Visual Studio: Fix paths to make them consistent between x86 and x64 archs #242

New in 1.1.6:

    * Migrate to GitHub Actions
    * Regenerate parsers with Ragel 7 and update build script
    * SAS7BDAT reader: Improved large file support on Windows #226
    * SAV reader: Skip null bytes in UTF-8 data https://github.com/tidyverse/haven/issues/560
    * SAV reader: Fix hang (oss-fuzz/23485)
    * DTA reader: Disallow str0 type
    * DTA reader: Fix encoding error when garbage values are present beyond the end of a string
    * Command file readers: Fix integer overflow (oss-fuzz/15778)
    * `extract_metadata`: Implement duration support #223 (thanks to @basgys)
    * Support for SAS files created with SAS Visual Forecaster #232
    * Report format widths for date/time SAS formats #233
    * Document the meaning of a -1 return value from `readstat_get_row_count` #234
    * Fix SAS file creation / modification times on Windows #238 #240

New in 1.1.5:

    * Support for building with MSVC #214 (thanks to @zebrys and @jonathon-love)
    * CLI tools: Support non-ASCII file paths on Windows #200 #216 (thanks to @zebrys)
    * DTA reader: Ignore bad timestamps
    * DTA writer: Fix memory leak
    * DTA writer: Improved support for empty value labels #219
    * POR reader: Improved support for date/time formats #160
    * SAS7BDAT reader: Added support for reading the dataset label #180 #213 (thanks to @reikoch)
    * SAS7BDAT reader: Improved detection of compressed files
    * SAS7BDAT reader: Improved bounds checking OSS-Fuzz/28312
    * SAS7BDAT reader: Support for more character encodings
    * SAV reader: Tolerate illegal lowercase variable names #217
    * SAV reader: Better support for non-UTF-8 variable names
    * SAV reader: Fix format widths for very long strings https://github.com/Roche/pyreadstat/issues/77
    * SAV reader: Fix undefined behavior with negative row counts OSS-Fuzz/23423

New in 1.1.4:

    * SAS7BDAT reader: Add support for binary-compressed files #21
    * XPT v8 writer: Improve compatibility with SAS #207 (thanks to @reikoch)
    * XPT reader: Fix reading of long variable names #208 (thanks to @reikoch) 
    * SAS readers: Support for more character encodings
    * SAV reader: Clients sometimes received truncated UTF-8 strings
    * SPSS writers: Improve compatibility with PSPP with DATETIME fields #211
    * All formats: Improved support for setting / getting the `display_width` #210

New in 1.1.3:

    * Fix warnings when compiling with GCC 10 #202
    * SAS RLE compressor: Fixes for large files #201
    * SAV reader: Improved support for UTF-8 column names #206
    * SAV reader: Return a better error message if the magic number doesn't match
    * SAV reader: Support for dash-separated timestamps

New in 1.1.2:

    * DTA reader: support for Spanish-locale timestamps
    * SAS reader: support for "any" encoding tidyverse/haven#482
    * CLI tool: Allow uppercase filename extensions
    * Improved support for reading SPSS and SAS command files
    * Improved support for reading POR files with format widths >100
    * Improved support for reading SAV files containing space-padded timestamps #197
    * Improved support for writing SAV files with a large number of variables #199
    * Improved support for reading SAS7BDAT files created by Stat/Transfer #198
    * Fix several integer overflows and undefined values #192 #193 #194 #195 #196

New in 1.1.1:

    * Support row limits in the plain-text parsers
    * SAV reader: Allow spaces in timestamp strings
    * README: Fix Windows / pacman instructions #189
    * Fix </sortlist> errors opening files in Stata 15 (tidyverse/haven#461)

New in 1.1:

    * New function: readstat_set_row_offset (#185). Thanks to @mikmart
    * Fix segfault when localtime fails on Windows
    * Fix implicit float conversion warning (oss-fuzz/16372)
    * New error code: READSTAT_ERROR_BAD_TIMESTAMP_VALUE
    * Renamed error code: READSTAT_ERROR_BAD_TIMESTAMP => READSTAT_ERROR_BAD_TIMESTAMP_STRING

New in 1.0.2:

    * Compilation: Fix -Wstringop-truncation warnings on GCC 8.2 and later (#151)
    * SPSS command parser: Fix signed integer overflow (oss-fuzz/15049)
    * POR parser: Use doubles internally to prevent integer overflows with very large exponents (#182)

New in 1.0.1:

    * SAV writer: Validate variable names
    * Fix a buffer overflow reading SPSS commands (oss-fuzz/15050) 
    * New error code READSTAT_ERROR_NAME_IS_ZERO_LENGTH when a blank variable name is provided
    * New fuzzing dictionary files in fuzz/dict for parsing plain-text file formats
    * Move corpus files from corpus to fuzz/corpus


src/readstat_dta.c
#include <stdlib.h>
#include <math.h>
#include <stdint.h>
#include <string.h>
#include <sys/types.h>

#include "readstat_dta.h"

#define DTA_MIN_VERSION 104
#define DTA_MAX_VERSION 118

dta_ctx_t *dta_ctx_alloc(readstat_io_t *io) {
    dta_ctx_t *ctx;
    if ((ctx = malloc(sizeof(dta_ctx_t))) == NULL) {
        return NULL;
    }
    memset(ctx, 0, sizeof(dta_ctx_t));

    ctx->io = io;
    ctx->initialized = 0;

    return ctx;
}

readstat_error_t dta_ctx_init(dta_ctx_t *ctx, int16_t nvar, int32_t nobs,
        unsigned char byteorder, unsigned char ds_format,
        const char *input_encoding, const char *output_encoding) {
    readstat_error_t retval = READSTAT_OK;
    int machine_byteorder = DTA_HILO;
    if (ds_format < DTA_MIN_VERSION || ds_format > DTA_MAX_VERSION)
        return READSTAT_ERROR_UNSUPPORTED_FILE_FORMAT_VERSION;

    if (machine_is_little_endian()) {
        machine_byteorder = DTA_LOHI;
    }

    if (byteorder != machine_byteorder) {
        ctx->machine_needs_byte_swap = 1;
    }

    ctx->nvar = ctx->machine_needs_byte_swap ? byteswap2(nvar) : nvar;
    ctx->nobs = ctx->machine_needs_byte_swap ? byteswap4(nobs) : nobs;
    
    ctx->machine_is_twos_complement = READSTAT_MACHINE_IS_TWOS_COMPLEMENT;

    if (ds_format < 105) {
        ctx->fmtlist_entry_len = 7;
    } else if (ds_format < 114) {
        ctx->fmtlist_entry_len = 12;
    } else if (ds_format < 118) {
        ctx->fmtlist_entry_len = 49;
    } else {
        ctx->fmtlist_entry_len = 57;
    }
    
    if (ds_format >= 117) {
        ctx->typlist_version = 117;
    } else if (ds_format >= 111) {
        ctx->typlist_version = 111;
    } else {
        ctx->typlist_version = 0;
    }

    if (ds_format >= 118) {
        ctx->data_label_len_len = 2;
    } else if (ds_format >= 117) {
        ctx->data_label_len_len = 1;
    }

    if (ds_format < 105) {
        ctx->expansion_len_len = 0;
    } else if (ds_format < 110) {
        ctx->expansion_len_len = 2;
    } else {
        ctx->expansion_len_len = 4;
    }
    
    if (ds_format < 110) {
        ctx->lbllist_entry_len = 9;
        ctx->variable_name_len = 9;
    } else if (ds_format < 118) {
        ctx->lbllist_entry_len = 33;
        ctx->variable_name_len = 33;
    } else {
        ctx->lbllist_entry_len = 129;
        ctx->variable_name_len = 129;
    }

    if (ds_format < 108) {
        ctx->variable_labels_entry_len = 32;
        ctx->data_label_len = 32;
    } else if (ds_format < 118) {
        ctx->variable_labels_entry_len = 81;
        ctx->data_label_len = 81;
    } else {
        ctx->variable_labels_entry_len = 321;
        ctx->data_label_len = 321;
    }

    if (ds_format < 105) {
        ctx->timestamp_len = 0;
        ctx->value_label_table_len_len = 2;
        ctx->value_label_table_labname_len = 12;
        ctx->value_label_table_padding_len = 2;
    } else {
        ctx->timestamp_len = 18;
        ctx->value_label_table_len_len = 4;
        if (ds_format < 118) {
            ctx->value_label_table_labname_len = 33;
        } else {
            ctx->value_label_table_labname_len = 129;
        }
        ctx->value_label_table_padding_len = 3;
    }

    if (ds_format < 117) {
        ctx->typlist_entry_len = 1;
        ctx->file_is_xmlish = 0;
    } else {
        ctx->typlist_entry_len = 2;
        ctx->file_is_xmlish = 1;
    }

    if (ds_format < 113) {
        ctx->max_char = DTA_OLD_MAX_CHAR;
        ctx->max_int16 = DTA_OLD_MAX_INT16;
        ctx->max_int32 = DTA_OLD_MAX_INT32;
        ctx->max_float = DTA_OLD_MAX_FLOAT;
        ctx->max_double = DTA_OLD_MAX_DOUBLE;
    } else {
        ctx->max_char = DTA_113_MAX_CHAR;
        ctx->max_int16 = DTA_113_MAX_INT16;
        ctx->max_int32 = DTA_113_MAX_INT32;
        ctx->max_float = DTA_113_MAX_FLOAT;
        ctx->max_double = DTA_113_MAX_DOUBLE;

        ctx->supports_tagged_missing = 1;
    }

    if (output_encoding) {
        if (input_encoding) {
            ctx->converter = iconv_open(output_encoding, input_encoding);
        } else if (ds_format < 118) {
            ctx->converter = iconv_open(output_encoding, "WINDOWS-1252");
        } else if (strcmp(output_encoding, "UTF-8") != 0) {
            ctx->converter = iconv_open(output_encoding, "UTF-8");
        }
        if (ctx->converter == (iconv_t)-1) {
            ctx->converter = NULL;
            retval = READSTAT_ERROR_UNSUPPORTED_CHARSET;
            goto cleanup;
        }
    }

    ctx->typlist_len = ctx->nvar * sizeof(uint16_t);
    ctx->varlist_len = ctx->variable_name_len * ctx->nvar * sizeof(char);
    ctx->srtlist_len = (ctx->nvar + 1) * sizeof(int16_t);
    ctx->fmtlist_len = ctx->fmtlist_entry_len * ctx->nvar * sizeof(char);
    ctx->lbllist_len = ctx->lbllist_entry_len * ctx->nvar * sizeof(char);
    ctx->variable_labels_len = ctx->variable_labels_entry_len * ctx->nvar * sizeof(char);

    if ((ctx->typlist = malloc(ctx->typlist_len)) == NULL) {
        retval = READSTAT_ERROR_MALLOC;
        goto cleanup;
    }
    if ((ctx->varlist = malloc(ctx->varlist_len)) == NULL) {
        retval = READSTAT_ERROR_MALLOC;
        goto cleanup;
    }
    if ((ctx->srtlist = malloc(ctx->srtlist_len)) == NULL) {
        retval = READSTAT_ERROR_MALLOC;
        goto cleanup;
    }
    if ((ctx->fmtlist = malloc(ctx->fmtlist_len)) == NULL) {
        retval = READSTAT_ERROR_MALLOC;
        goto cleanup;
    }
    if ((ctx->lbllist = malloc(ctx->lbllist_len)) == NULL) {
        retval = READSTAT_ERROR_MALLOC;
        goto cleanup;
    }

    if ((ctx->variable_labels = malloc(ctx->variable_labels_len)) == NULL) {
        retval = READSTAT_ERROR_MALLOC;
        goto cleanup;
    }

    ctx->initialized = 1;

cleanup:
    return retval;
}

void dta_ctx_free(dta_ctx_t *ctx) {
    if (ctx->typlist)
        free(ctx->typlist);
    if (ctx->varlist)
        free(ctx->varlist);
    if (ctx->srtlist)
        free(ctx->srtlist);
    if (ctx->fmtlist)
        free(ctx->fmtlist);
    if (ctx->lbllist)
        free(ctx->lbllist);
    if (ctx->variable_labels)
        free(ctx->variable_labels);
    if (ctx->converter)
        iconv_close(ctx->converter);
    if (ctx->data_label)
        free(ctx->data_label);
    free(ctx);
}

readstat_types_t dta_type_info(uint16_t typecode, size_t *max_len, dta_ctx_t *ctx) {
    size_t len = 0;
    readstat_types_t type = READSTAT_TYPE_STRING;
    if (ctx->typlist_version == 111) {
        switch (typecode) {
            case DTA_111_TYPE_CODE_CHAR:
                len = 1; type = READSTAT_TYPE_CHAR; break;
            case DTA_111_TYPE_CODE_INT16:
                len = 2; type = READSTAT_TYPE_INT16; break;
            case DTA_111_TYPE_CODE_INT32:
                len = 4; type = READSTAT_TYPE_INT32; break;
            case DTA_111_TYPE_CODE_FLOAT:
                len = 4; type = READSTAT_TYPE_FLOAT; break;
            case DTA_111_TYPE_CODE_DOUBLE:
                len = 8; type = READSTAT_TYPE_DOUBLE; break;
            default:
                len = typecode; type = READSTAT_TYPE_STRING; break;
        }
    } else if (ctx->typlist_version == 117) {
        switch (typecode) {
            case DTA_117_TYPE_CODE_CHAR:
                len = 1; type = READSTAT_TYPE_CHAR; break;
            case DTA_117_TYPE_CODE_INT16:
                len = 2; type = READSTAT_TYPE_INT16; break;
            case DTA_117_TYPE_CODE_INT32:
                len = 4; type = READSTAT_TYPE_INT32; break;
            case DTA_117_TYPE_CODE_FLOAT:
                len = 4; type = READSTAT_TYPE_FLOAT; break;
            case DTA_117_TYPE_CODE_DOUBLE:
                len = 8; type = READSTAT_TYPE_DOUBLE; break;
            case DTA_117_TYPE_CODE_STRL:
                len = 8; type = READSTAT_TYPE_LONG_STRING; break;
            default:
                len = typecode; type = READSTAT_TYPE_STRING; break;
        }
    } else {
        switch (typecode) {
            case DTA_OLD_TYPE_CODE_CHAR:
                len = 1; type = READSTAT_TYPE_CHAR; break;
            case DTA_OLD_TYPE_CODE_INT16:
                len = 2; type = READSTAT_TYPE_INT16; break;
            case DTA_OLD_TYPE_CODE_INT32:
                len = 4; type = READSTAT_TYPE_INT32; break;
            case DTA_OLD_TYPE_CODE_FLOAT:
                len = 4; type = READSTAT_TYPE_FLOAT; break;
            case DTA_OLD_TYPE_CODE_DOUBLE:
                len = 8; type = READSTAT_TYPE_DOUBLE; break;
            default:
                len = typecode - 0x7F; type = READSTAT_TYPE_STRING; break;
        }
    }
    
    *max_len = len;
    return type;
}

